### ISSUE TITLE
Refactor generate-coder-table into module

### ISSUE THREAD

**Issue by davidskalinder** [davidskalinder/mpeds-coder#132]
All right, so, new plan for this: I'm going to refactor what's in `generate-coder-table.py` and put it into [a subfolder](https://github.com/davidskalinder/mpeds-coder/issues/127) so it can be run by either the UI controller or a shell user.

Also, here's a strategy that I think might work for the timeout errors:

- Write, e.g., `count_records()` as a separate function
- Take the code that writes a single chunk out of the loop and put it in, e.g., `write_chunk()`
- Have `generateCoderAudit()` or a wrapper in the module call `write_chunk()` until it's done

It's not entirely clear to me how to correctly handle all those requests.  But I'm hoping refactoring `generate-coder-table.py` into distinct methods will help clarify things.

_Originally posted by @davidskalinder in https://github.com/davidskalinder/mpeds-coder/issues/122#issuecomment-850585674_

**Comment by davidskalinder** [comment]
From #122: 

> At adaa586, set up a working test of the module with nothing in it. (Python's modules take some re-getting used to!)

> NB no need for a class just yet I don't think: bare functions in the module seem fine until there's a need for things to have states?

**Comment by davidskalinder** [comment]
Am working on this kind of slowly and stupidly.  ace3d79cf9cfdb technically works by just-barely wrapping the existing functionality in a call from the controller.

The code should really be split though.  It seems like it shouldn't be too hard to have all the chunk-writing code be shared for several exports and then have a separate function that just handles the query logic for each one.

**Comment by davidskalinder** [comment]
Hey @alexhanna I notice that the current coder-table generator leaves the dashes in publication names (like `Tufts-Daily:-Tufts-University`).

Does anything depend on those dashes being there?  Now that we have the publication names in the DB, we could just get them from there rather than chiseling them out of the `db_id` string; but then we'd get the names prior to the spaces being replaced by hyphens.

**Comment by alexhanna** [comment]
That's true. But yes, I actually think some things depend on them. I have
to dig back into the ETL and mpeds_coder.py code to see what, though.

**Comment by davidskalinder** [comment]
Hmm okay maybe for now it's best to keep on chiseling then

**Comment by davidskalinder** [comment]
Wow, SQLAlchemy really hates returning query results of model classes as columns and query results of columns as classes, huh.  This makes it super-annoying to do something like make a function that handles an arbitrary export result, since the columns of the result will be at different depths depending on how the query was built (and I can't find a way to flatten the results).

Not sure how to handle this.  I suspect that the real solution is to make a model class that joins all the fields we want in the export and then just have the export writer query that model; but I'm darned if I can find any comprehensible docs on how to make a multi-table model in Flask.

One alternative is to always query whole models to ensure that everything's at the same depth and then use a trick like [this](https://stackoverflow.com/questions/9486180/sql-alchemy-orm-returning-a-single-column-how-to-avoid-common-post-processing) to iterate through everything.  The downside is that then the full article texts get loaded all the time; so at 75b95d1 I tried to set those to only load when needed (though I can find no documentation on how to test deferred loading, so hey, who knows what will happen!).

The only other possibility I can think of is to have separate export-writers that each know the exact structure of what they're writing.  Hopefully if it comes to that we could still abstract some of the chunking code at least.  But man that'd be annoying.

The more documentation I read for Flask and SQLAlchemy, the more I appreciate the Tidyverse.

**Comment by davidskalinder** [comment]
> One alternative is to always query whole models to ensure that everything's at the same depth

I thought of another similar alternative, which is to always query individual columns.  This might be better since I think one could get the column names of the model using some kind of trick with [`__table__`](https://stackoverflow.com/questions/6455560/how-to-get-column-names-from-sqlalchemy-result-declarative-syntax) or some such; so that might make it easy to always get all the column names into a columnwise query that can then be passed to the writer.

I suspect this will be the best solution until/unless we figure out how to store classes of data corresponding to domain objects rather than data objects?

**Comment by davidskalinder** [comment]
> I think one could get the column names of the model using some kind of trick with [`__table__`](https://stackoverflow.com/questions/6455560/how-to-get-column-names-from-sqlalchemy-result-declarative-syntax) or some such

Well after several hours of hunting I can't find this hypothesized trick.  I can get the column names and I can get the table names, but that doesn't help since SQLAlchemy needs columns to be specified with the *class* name. Like maaybe we could go the route of class introspection to do this... but then that will require some further kludges to get it into a form that can be used for column specs.

On the whole, this task continues to feel like fighting through a pool of molasses with a machete.  This signals to me that I fundamentally don't understand something about how SQLAlchemy is supposed to work in cases like these (or of any other way to specify a dataset with domain logic), and I don't know what to do about it.

I'm also getting worried that this task might not be achievable by handover time.  At some stage soon I'll have to press the stop-loss button and write up recipes for the existing export methods (of which I believe some are broken).
